src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java

```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
      *          be found.
      */
     public TypeMirror getFirstVisibleSuperClass(TypeMirror type) {
-        List<? extends TypeMirror> superTypes = typeUtils.directSupertypes(type);
-        TypeMirror superType = superTypes.isEmpty() ? getObjectType() : superTypes.get(0);
-        TypeElement superClass = asTypeElement(superType);
-        // skip "hidden" classes
-        while ((superClass != null && hasHiddenTag(superClass))
-                || (superClass != null &&  !isPublic(superClass) && !isLinkable(superClass))) {
-            TypeMirror supersuperType = superClass.getSuperclass();
-            TypeElement supersuperClass = asTypeElement(supersuperType);
-            if (supersuperClass == null
-                    || supersuperClass.getQualifiedName().equals(superClass.getQualifiedName())) {
-                break;
-            }
-            superType = supersuperType;
-            superClass = supersuperClass;
-        }
-        if (typeUtils.isSameType(type, superType)) {
-            return null;
+        // TODO: this computation should be eventually delegated to VisibleMemberTable
+        Set<TypeElement> alreadySeen = null;
+        assert (alreadySeen = new HashSet<>()) != null; // create set conditionally
+        for (var t = type; ;) {
+            var supertypes = typeUtils.directSupertypes(t);
+            if (supertypes.isEmpty()) { // end of hierarchy
+                return null;
+            }
+            t = supertypes.get(0); // if non-empty, the first element is always the superclass
+            var te = asTypeElement(t);
+            assert alreadySeen.add(te);
```

Comment:
I guess it's OK, but it is a mild uugh to see an imperative method as an expression for `assert`